api/src/pcapi/admin/custom_views/beneficiary_user_view.py

```diff
@@ -112,6 +119,50 @@ def beneficiary_digital_remaining_formatter(view, context, model, name) -> Marku
     return Markup("<span>{digital_remaining}&nbsp;&euro;</span>").format(digital_remaining=digital_remaining)
 
 
+def _has_underage_deposit(user: User) -> bool:
+    return user.deposit and user.deposit.type == finance_models.DepositType.GRANT_15_17
+
+
+def _update_underage_beneficiary_deposit_expiration_date(user: User) -> None:
+    if user.birth_date is None:
+        raise ValueError("User has no birth_date")
+
+    current_deposit_expiration_date = user.deposit.expirationDate
+    user_eighteenth_birthday = datetime.datetime.combine(user.birth_date + relativedelta(years=18), datetime.time.min)
+
+    if current_deposit_expiration_date == user_eighteenth_birthday:
+        return
+
+    logger.info(
+        "Updating deposit expiration date for underage beneficiary %s",
+        user.id,
+        extra={
+            "user": user.id,
+            "deposit": user.deposit.id,
+            "current_expiration_date": current_deposit_expiration_date,
+            "new_expiration_date": user_eighteenth_birthday,
+        },
+    )
+
+    # If both dates are in the past, we don't update the expiration date (deposit is already expired)
+    if (
+        current_deposit_expiration_date < datetime.datetime.utcnow()
+        and user_eighteenth_birthday < datetime.datetime.utcnow()
+    ):
+        return
+
+    # If the current expiration date is before the user's 18th birthday, we update it
+    if current_deposit_expiration_date < user_eighteenth_birthday:
+        user.deposit.expirationDate = user_eighteenth_birthday
+
+    # If the current expiration date is after the user's 18th birthday, we set it to the user's 18th birthday if it's in the future, else immediately
+    # That way, the deposit will expire and not mess with bookings that may have happened after the user's 18th birthday.
+    elif current_deposit_expiration_date > user_eighteenth_birthday:
+        user.deposit.expirationDate = max(user_eighteenth_birthday, datetime.datetime.utcnow())
```

Comment:
Ce serait plus facile à comprendre si on distinguait les cas : 
```
if user_eighteenth_birthday > datetime.now():
   user.deposit.expirationDate = user_eighteenth_birthday
else:
   date = now()
````

Mais encore mieux, je vois qu'on peut simplifier le tout : 
```
if eighteen_birthday > now(): 
   # extend deposit expiration date
   expirationDate = eighteen_birthday
else if prevDate < now():
       # no need to update the deposit expirationDate because it is already passed
     return
else:
   # reduce to now and not to theoreticalNewDate in case there are bookings made between these dates
   expirationDate = now()
```